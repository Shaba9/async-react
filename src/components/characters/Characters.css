ul li {
  list-style: none;
  display: inline-block;
  margin: 2vw;
}