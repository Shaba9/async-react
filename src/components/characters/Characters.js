import React, { PureComponent } from 'react';
import Character from './Character';
import Paging from '../paging/Paging';
import { getCharacters } from '../../services/rickAndMortyApi';
import styles from './Characters.css';

export default class Characters extends PureComponent {
  state = {
    characters: [],
    currentPage: 1,
    totalPages: 1
  }

  fetch = page => {
    getCharacters(page)
      .then(res => {
        this.setState({ characters: res.results, totalPages: res.totalPages });
      });
  }

  componentDidMount() {
    this.fetch(1);
  }

  componentDidUpdate(prevProps, prevState) {
    const { currentPage } = this.state;
    if(prevState.currentPage !== currentPage) {
      this.fetch(currentPage);
    }
  }

  increment = () => {
    let page = this.state.currentPage;

    page === this.state.totalPages ? page = 1 : page++;
    this.setState({ currentPage: page });
  }

  decrement = () => {
    let page = this.state.currentPage;

    page === 1 ? page = this.state.totalPages : page--;
    this.setState({ currentPage: page });
  }

  render() {
    const charactersList = this.state.characters.map(char => {
      return <Character
        key={char.id}
        character={char}
      />;
    });

    const { currentPage, totalPages } = this.state;

    return (
      <div className={styles.characters}>
        <Paging
          currentPage={currentPage}
          totalPages={totalPages}
          decrement={this.decrement}
          increment={this.increment}
        />
        <ul>
          {charactersList}
        </ul>
      </div>
    );
  }
}
