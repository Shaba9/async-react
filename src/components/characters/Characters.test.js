import React from 'react';
import renderer from 'react-test-renderer';
import Characters from './Characters';

describe('Character', () => {
  it('matches a snapshot', () => {
    const characters = [
      {
        name: 'bob',
        img: 'https://www.littledayout.com/wp-content/uploads/articles/2017/07-July/20170715-WorldEmojiDay/DabEmoji.jpg',
        status: 'alive',
        species: 'human',
        gender: 'mail'
      },
      {
        name: 'bob2',
        img: 'https://www.littledayout.com/wp-content/uploads/articles/2017/07-July/20170715-WorldEmojiDay/DabEmoji.jpg',
        status: 'dead',
        species: 'animal',
        gender: 'mail' 
      }
    ];

    const tree = renderer.create(
      <Characters
        characters={characters}
      />
    ).toJSON();
  
    expect(tree).toMatchSnapshot();
  });
});
