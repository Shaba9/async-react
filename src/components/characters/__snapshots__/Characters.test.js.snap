// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character matches a snapshot 1`] = `
<div
  className="characters"
>
  <header>
    <p>
      Current Page: 
      1
    </p>
    <p>
      Total Pages: 
      1
    </p>
  </header>
  <button
    onClick={[Function]}
  >
    &lt;
  </button>
  <button
    onClick={[Function]}
  >
    &gt;
  </button>
  <ul />
</div>
`;
